crates/ty_python_semantic/src/types/infer.rs

```diff
@@ -6702,13 +6702,24 @@ impl<'db, 'ast> TypeInferenceBuilder<'db, 'ast> {
                 Some(Type::BooleanLiteral(b1 ^ b2))
             }
 
-            (Type::BooleanLiteral(bool_value), right, op) => self.infer_binary_expression_type(
-                node,
-                emitted_division_by_zero_diagnostic,
-                Type::IntLiteral(i64::from(bool_value)),
-                right,
-                op,
-            ),
+            (Type::BooleanLiteral(bool_value), right, op) => {
+                if let Some(nominal_instance) = right.into_nominal_instance() {
+                    let class_literal = match nominal_instance.class {
+                        ClassType::NonGeneric(class_literal) => class_literal,
+                        ClassType::Generic(generic_alias) => generic_alias.origin(self.db()),
+                    };
+                    if class_literal.is_known(self.db(), KnownClass::Bool) {
+                        return Some(right);
```

Comment:
I don't think we can just assume the output is a bool without even knowing what operation is being performed -- the better approach here would be to recursively call `self.infer_binary_expression_type` but with the left operand replaced by `KnownClass::Bool.to_instance()`. Similar to how we currently fall back to `Type::IntLiteral(...)`